### PR TITLE
Yqxu dev

### PR DESCRIPTION
**Pull request by yqxu**
**What this PR does**:
using tire to match url instead of avgtree

### CODE REVIEW

**Comment by cityiron** [comment]
I plan this approve to 0.3 or 0.4 version, may be in 2020.03

File: pkg/router/route.go [hunk 1]
```diff
@@ -114,27 +107,36 @@ func (rt *Route) findNode(fullPath string) (*Node, bool) {
...
 	var found bool
-	if n, found = rt.searchWildcard(lowerPath); !found {
-		rt.lock.RLock()
-		defer rt.lock.RUnlock()
-		if n, found = rt.tree.Get(lowerPath); !found {
-			return nil, false
-		}
+	rt.lock.RLock()
+	defer rt.lock.RUnlock()
+	tireNode, _, _ := rt.tree.Get(lowerPath)
```
Comment (williamfeng323) [inline pkg/router/route.go:112]: please update the variable name to trieNode

File: pkg/router/route.go [hunk 2]
```diff
@@ -114,27 +107,36 @@ func (rt *Route) findNode(fullPath string) (*Node, bool) {
...
+	var found bool
 	rt.lock.RLock()
 	defer rt.lock.RUnlock()
-	wildcardPaths := rt.wildcardTree.Keys()
-	for _, p := range wildcardPaths {
-		if wildcardMatch(p.(string), fullPath) != nil {
-			n, ok := rt.wildcardTree.Get(p)
-			return n.(*Node), ok
-		}
+	tireNode, _, _ := rt.tree.Match(lowerPath)
```
Comment (williamfeng323) [inline pkg/router/route.go:130]: please help update the variable to trieNode

File: pkg/router/route.go [hunk 3]
```diff
@@ -114,27 +107,36 @@ func (rt *Route) findNode(fullPath string) (*Node, bool) {
...
+	}
+	n = tireNode.GetBizInfo()
+	if n == nil {
+		return nil, false
 	}
 	return n.(*Node), found
 }
 
-func (rt *Route) searchWildcard(fullPath string) (*Node, bool) {
+func (rt *Route) matchNode(fullPath string) (*Node, bool) {
```
Comment (williamfeng323) [inline pkg/router/route.go:124]: it seems this function is the same as findNode?

File: pkg/utils/trie/trie.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+type Node struct {
+	matchStr         string           //冗余信息 通配节点冗余变量名，普通节点冗余节点名
+	children         map[string]*Node //子树
+	PathVariablesSet map[string]*Node //变量名集合 找不到set先用map todo
+	PathVariableNode *Node            //通配变量节点后的子树
+	endOfPath        bool             //是否是路径末尾
+	bizInfo          interface{}      //随便塞啥
+}
+
+//Tire对外暴露 推荐外部使用
```
Comment (williamfeng323) [inline pkg/utils/trie/trie.go:44]: please correct the spelling to Trie

File: pkg/utils/trie/trie.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+//Put put key and values into trie as map.
+func (trie *Trie) Put(withOutHost string, bizInfo interface{}) bool {
+	parts := urlpath.Split(withOutHost)
```
Comment (williamfeng323) [inline pkg/utils/trie/trie.go:48]: should be paths?

File: pkg/utils/trie/trie.go (hunk 4 above)
Comment (AlexStocks) [reply pkg/utils/trie/trie.go:44]: Pls do not use chinese comment.

File: pkg/router/route.go (hunk 3 above)
Comment (yqxu) [reply pkg/router/route.go:124]: find Node only traversing the tree, ignore  path variable  :=var . 
'/test/test/xxx'  can match '/test/test/:a'  , but '/test/test/xxx' can't find '/test/test/:a'

**Pull request closed without merging**